src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/IncludeLocalesPlugin.java

```diff
@@ -158,10 +157,9 @@ public ResourcePool transform(ResourcePool in, ResourcePoolBuilder out) {
                 if (resource != null &&
                     resource.type().equals(ResourcePoolEntry.Type.CLASS_OR_RESOURCE)) {
                     byte[] bytes = resource.contentBytes();
-                    ClassReader cr = newClassReader(path, bytes);
-                    if (Arrays.stream(cr.getInterfaces())
-                        .anyMatch(i -> i.contains(METAINFONAME)) &&
-                        stripUnsupportedLocales(bytes, cr)) {
+                    if (newClassReader(path, bytes).interfaces().stream()
```

Comment:
`IncludeLocalesPluginTest.java` test is failing.   Adding `-J-Djlink.debug=true` will get the stack trace:
 
```
Failed to parse class file: /jdk.localedata/sun/text/resources/ext/LineBreakIteratorData_th
java.lang.IllegalStateException: Bad magic number
	at java.base/jdk.internal.classfile.impl.ClassReaderImpl.<init>(ClassReaderImpl.java:94)
	at java.base/jdk.internal.classfile.impl.ClassImpl.<init>(ClassImpl.java:68)
	at java.base/jdk.internal.classfile.Classfile.parse(Classfile.java:165)
	at jdk.jlink/jdk.tools.jlink.internal.plugins.AbstractPlugin.newClassReader(AbstractPlugin.java:105)
	at jdk.jlink/jdk.tools.jlink.internal.plugins.IncludeLocalesPlugin.lambda$transform$5(IncludeLocalesPlugin.java:161)
	at jdk.jlink/jdk.tools.jlink.plugin.ResourcePool.lambda$transformAndCopy$0(ResourcePool.java:112)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1924)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at jdk.jlink/jdk.tools.jlink.plugin.ResourcePool.transformAndCopy(ResourcePool.java:111)
	at jdk.jlink/jdk.tools.jlink.internal.plugins.IncludeLocalesPlugin.transform(IncludeLocalesPlugin.java:154)
	at jdk.jlink/jdk.tools.jlink.internal.ImagePluginStack.visitResources(ImagePluginStack.java:262)
	at jdk.jlink/jdk.tools.jlink.internal.ImageFileCreator.generateJImage(ImageFileCreator.java:184)
	at jdk.jlink/jdk.tools.jlink.internal.ImageFileCreator.writeImage(ImageFileCreator.java:163)
	at jdk.jlink/jdk.tools.jlink.internal.ImageFileCreator.create(ImageFileCreator.java:100)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask$ImageHelper.retrieve(JlinkTask.java:889)
	at jdk.jlink/jdk.tools.jlink.internal.ImagePluginStack.operate(ImagePluginStack.java:194)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.createImage(JlinkTask.java:452)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.run(JlinkTask.java:292)
	at jdk.jlink/jdk.tools.jlink.internal.Main.run(Main.java:56)
	at jdk.jlink/jdk.tools.jlink.internal.Main.main(Main.java:34)
Dumping class file 26549-include-locales//jdk.localedata/sun/text/resources/ext/LineBreakIteratorData_th
java.lang.IllegalStateException: Bad magic number
	at java.base/jdk.internal.classfile.impl.ClassReaderImpl.<init>(ClassReaderImpl.java:94)
	at java.base/jdk.internal.classfile.impl.ClassImpl.<init>(ClassImpl.java:68)
	at java.base/jdk.internal.classfile.Classfile.parse(Classfile.java:165)
	at jdk.jlink/jdk.tools.jlink.internal.plugins.AbstractPlugin.newClassReader(AbstractPlugin.java:105)
	at jdk.jlink/jdk.tools.jlink.internal.plugins.IncludeLocalesPlugin.lambda$transform$5(IncludeLocalesPlugin.java:161)
	at jdk.jlink/jdk.tools.jlink.plugin.ResourcePool.lambda$transformAndCopy$0(ResourcePool.java:112)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1924)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at jdk.jlink/jdk.tools.jlink.plugin.ResourcePool.transformAndCopy(ResourcePool.java:111)
	at jdk.jlink/jdk.tools.jlink.internal.plugins.IncludeLocalesPlugin.transform(IncludeLocalesPlugin.java:154)
	at jdk.jlink/jdk.tools.jlink.internal.ImagePluginStack.visitResources(ImagePluginStack.java:262)
	at jdk.jlink/jdk.tools.jlink.internal.ImageFileCreator.generateJImage(ImageFileCreator.java:184)
	at jdk.jlink/jdk.tools.jlink.internal.ImageFileCreator.writeImage(ImageFileCreator.java:163)
	at jdk.jlink/jdk.tools.jlink.internal.ImageFileCreator.create(ImageFileCreator.java:100)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask$ImageHelper.retrieve(JlinkTask.java:889)
	at jdk.jlink/jdk.tools.jlink.internal.ImagePluginStack.operate(ImagePluginStack.java:194)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.createImage(JlinkTask.java:452)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.run(JlinkTask.java:292)
	at jdk.jlink/jdk.tools.jlink.internal.Main.run(Main.java:56)
	at jdk.jlink/jdk.tools.jlink.internal.Main.main(Main.java:34)
```	

Apparently `/jdk.localedata/sun/text/resources/ext/LineBreakIteratorData_th` is not a valid class file.  Using ASM `ClassReader` to parse this resource file succeeds because it does not do much of the class file validation.    Instead the ClassFile API throws `IllegalStateException` as it has a bad magic number.

@naotoj what is the file format of this `LineBreakIteratorData_th` file?

Reply:
This plugin does not fit to the Classfile API use cases portfolio at all. Here it is only needed to get interfaces list out of the constant pool and then the Utf8 CP entries are in-place modified. That is out of the Classfile API or ASM capabilities. Bad magic number is not the only difference, both libraries fail whenever touching even headers of any other standard classfile elements, only constant pool and interfaces seem to be "safe". 
I can prepare very simple standalone helper class dealing with this case to remove ASM dependency. 
However knowing more about the file format is always useful :)

Reply:
The file `LineBreakIteratorData_th` is not a class file, thus the exception is appropriate. This plugin test is trying to verify the integrity of the generated jimage, not only the required class files for the specified locale, but also the required resources as such. The file itself is a binary file that the BreakIterator uses internally.

Reply:
@naotoj as these localedata resource files are not `.class`, the plugin should not use ASM (or Classfile API) to parse it but ASM ClassReader just happens to work.   Is the format documented anywhere?

@asotona  I agree that this ASM dependency should be replaced with a simple class to parse the localedata resources.   We can create a follow-up issue to document or validate the file format if appropriate.

Reply:
There is no common binary format for "localedata", the file in question is the custom format specific for `BreakIterator`, ancient code from Taligent. IIUC, only description of the format is in the source: https://github.com/openjdk/jdk/blob/2f23c80e0de44815d26a7d541701e16c9c1d32bc/src/java.base/share/classes/sun/text/RuleBasedBreakIterator.java#L325

Reply:
I further clarified with Naoto and I now understand the problem.    The plugin intends to transform `.class` that implements `LocaleDataMetaInfo` in `jdk.localedata` module.   It's a bug in the plugin `--include-locales=*` which will read all resources from `jdk.localedata` with ASM ClassReader and the plugin just happens to work.  The plugin should only read `.class` entries or specific `LocaleDataMetaInfo` implementation classes.    I created https://bugs.openjdk.org/browse/JDK-8304367 that Naoto will help fixing. 

@asotona your previous version converting it to ClassFile API should work after JDK-8304367.

Reply:
Submitted: https://github.com/openjdk/jdk/pull/13067

Reply:
I've reverted `IncludeLocalesPlugin` to previous version, plus minor fixes and together with #13067 the `IncludeLocalesPluginTest` is passing.
Thank you.

Reply:
The fix to `IncludeLocalesPlugin` has been integrated. @asotona would you please check that the exception is not thrown with your changes?

Reply:
Yes, I can confirm the test now passes.

Reply:
Great! Thanks for checking